zone/raids.cpp

```diff
@@ -2203,3 +2257,734 @@ void Raid::SetNewRaidLeader(uint32 i)
+		MarkedUpdate{.mark_target = RaidMarkTarget2, .slot = MAIN_MARKER_2_SLOT},
+		MarkedUpdate{.mark_target = RaidMarkTarget3, .slot = MAIN_MARKER_3_SLOT},
+	};
+
+	for (auto& u : marked_updates) {
+		if (marked_npcs[u.slot]) {
+			auto m = entity_list.GetMob(marked_npcs[u.slot]);
+			if (m && m->GetHP() > 0) {
+				UpdateXTargetType(u.mark_target, m, m->GetName());
+			}
+			else {
+				UpdateXTargetType(u.mark_target, nullptr);
+			}
+		}
+		else {
+			UpdateXTargetType(u.mark_target, nullptr);
+		}
+	}
+}
+
+void Raid::DelegateAbilityMark(Mob* delegator, const char* delegatee)
+{
+	auto raid_delegatee = entity_list.GetRaidByName(delegatee);
+	if (!raid_delegatee) {
+		delegator->CastToClient()->MessageString(Chat::Cyan, 5082, delegatee);
+		return;
+	}
+	uint32 raid_delegatee_id = raid_delegatee->GetID();
+	uint32 raid_delegator_id = GetID();
+	if (raid_delegatee_id != raid_delegator_id) {
+		delegator->CastToClient()->MessageString(Chat::Cyan, NOT_IN_YOUR_RAID, delegatee);
+		return;
+	}
+
+	auto rm = &members[GetPlayerIndex(delegatee)];
+	if (!rm) {
+		return;
+	}
+	auto c = rm->member;
+	if (!c) {
+		return;
+	}
+
+	auto slot = FindNextRaidDelegateSlot(FindNextMarkerSlot);
+	auto mm = rm->main_marker;
+
+	if (slot == -1 && !mm) {
+		delegator->CastToClient()->MessageString(Chat::Cyan, MAX_MAIN_RAID_MARKERS);
+		return;
+	}
+
+	auto outapp = new EQApplicationPacket(OP_RaidDelegateAbility, sizeof(DelegateAbility_Struct));
+	DelegateAbility_Struct* das = (DelegateAbility_Struct*)outapp->pBuffer;
+	if (mm) {
+		das->Action = ClearDelegate;
+		memset(main_marker_pcs[mm - 1], 0, 64);
+		rm->main_marker = 0;
+		std::string query = StringFormat("UPDATE raid_members SET is_marker = 0 WHERE raidid = %i AND name= '%s';",
+			GetID(),
+			delegatee
+		);
+		auto results = database.QueryDatabase(query);
+		if (!results.Success()) {
+			LogError("Unable to clear raid main marker for player: [{}] with error [{}]\n",
+				delegatee,
+				results.ErrorMessage().c_str()
+			);
+		}
+	}
+	else {
+		if (slot >= 0) {
+			strcpy(main_marker_pcs[slot], c->GetName());
+			rm->main_marker = slot + 1;
+			das->Action = SetDelegate;
+			std::string query = StringFormat("UPDATE raid_members SET is_marker = %i WHERE raidid = %i AND name= '%s';",
+				slot + 1,
+				this->GetID(),
+				delegatee
+			);
+			auto results = database.QueryDatabase(query);
+			if (!results.Success()) {
+				LogError("Unable to set raid main marker for player: [{}] slot: [{}] with error [{}]\n",
+					delegatee,
+					slot + 1,
+					results.ErrorMessage().c_str()
+				);
+			}
+		}
+	}
+	das->DelegateAbility = RaidDelegateMainMarker;
+	das->MemberNumber = 0;
+	das->EntityID = c->GetID();
+	strcpy(das->Name, delegatee);
+	QueuePacket(outapp);
+	safe_delete(outapp);
+}
+
+int Raid::FindNextRaidDelegateSlot(int option) 
+{
+	//option 1 FindNextRaidMainMarkerSlot
+	//option 2 FindNextRaidMainAssisterSlot
+	if (option == 1) {
```

Comment:
These should be constants (remember the magic numbers feedback)

Reply:
I should have noticed this, updated now.